### PR TITLE
Built-in endpoints history configuration <master> [7705]

### PR DESCRIPTION
**Pull request by IkerLuengo**
This is a port of #907 from 1.9.x

### CODE REVIEW

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1439/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1212/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1437/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1478/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1251/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1476/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1498/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1269/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1496/)

File: src/cpp/rtps/builtin/discovery/endpoint/EDPServer.cpp [hunk 1]
```diff
@@ -137,65 +137,75 @@ bool EDPServer::createSEDPEndpoints()
...
 
     std::copy_if(history.changesBegin(), history.changesBegin(), std::front_inserter(removal),
-        [_demises](const CacheChange_t* chan)
-        {
-            return _demises.find(chan->instanceHandle) != _demises.cend();
-        });
+            [_demises](const CacheChange_t* chan)
+                {
+                    return _demises.find(chan->instanceHandle) != _demises.cend();
+                });
```
Comment (IkerLuengo) [inline src/cpp/rtps/builtin/discovery/endpoint/EDPServer.cpp]:
```suggestion
        [_demises](const CacheChange_t* chan)
            {
                return _demises.find(chan->instanceHandle) != _demises.cend();
            });
```

File: src/cpp/rtps/builtin/discovery/endpoint/EDPServer.cpp [hunk 2]
```diff
@@ -206,22 +216,24 @@ bool EDPServer::trimWriterHistory(
...
     if (sid == SampleIdentity::unknown())
     {
         sid.writer_guid(c.writerGUID);
         sid.sequence_number(c.sequenceNumber);
-        logError(RTPS_EDP, "A DATA(r|w) received by server " << writer.getGuid()
-            << " from participant " << c.writerGUID << " without a valid SampleIdentity");
+        logError(RTPS_EDP,
+                "A DATA(r|w) received by server " << writer.getGuid()
+                                                  << " from participant " << c.writerGUID <<
+                            " without a valid SampleIdentity");
```
Comment (IkerLuengo) [inline src/cpp/rtps/builtin/discovery/endpoint/EDPServer.cpp]:
```suggestion
        logError(RTPS_EDP,
                "A DATA(r|w) received by server " << writer.getGuid()
                    << " from participant " << c.writerGUID
                    << " without a valid SampleIdentity");
```

File: src/cpp/rtps/builtin/discovery/endpoint/EDPServer.cpp [hunk 3]
```diff
@@ -286,11 +299,11 @@ bool EDPServer::removeLocalReader(RTPSReader* R)
...
-        CacheChange_t* change = writer->first->new_change([]() -> uint32_t
-            {
-                return DISCOVERY_SUBSCRIPTION_DATA_MAX_SIZE;
-            },
-            NOT_ALIVE_DISPOSED_UNREGISTERED, iH);
+        CacheChange_t* change = writer->first->new_change([this]() -> uint32_t
+                    {
+                        return mp_PDP->builtin_attributes().writerPayloadSize;
+                    },
+                        NOT_ALIVE_DISPOSED_UNREGISTERED, iH);
```
Comment (IkerLuengo) [inline src/cpp/rtps/builtin/discovery/endpoint/EDPServer.cpp]:
```suggestion
        CacheChange_t* change = writer->first->new_change(
            [this]() -> uint32_t
            {
                return mp_PDP->builtin_attributes().writerPayloadSize;
            },
            NOT_ALIVE_DISPOSED_UNREGISTERED, iH);
```

File: src/cpp/rtps/builtin/discovery/endpoint/EDPServer.cpp [hunk 4]
```diff
@@ -321,11 +335,11 @@ bool EDPServer::removeLocalWriter(RTPSWriter* W)
...
-        CacheChange_t* change = writer->first->new_change([]() -> uint32_t
-            {
-                return DISCOVERY_PUBLICATION_DATA_MAX_SIZE;
-            },
-            NOT_ALIVE_DISPOSED_UNREGISTERED, iH);
+        CacheChange_t* change = writer->first->new_change([this]() -> uint32_t
+                    {
+                        return mp_PDP->builtin_attributes().writerPayloadSize;
+                    },
+                        NOT_ALIVE_DISPOSED_UNREGISTERED, iH);
```
Comment (IkerLuengo) [inline src/cpp/rtps/builtin/discovery/endpoint/EDPServer.cpp]:
```suggestion
        CacheChange_t* change = writer->first->new_change(
            [this]() -> uint32_t
            {
                return mp_PDP->builtin_attributes().writerPayloadSize;
            },
            NOT_ALIVE_DISPOSED_UNREGISTERED, iH);
```

File: src/cpp/rtps/builtin/discovery/endpoint/EDPServer.cpp [hunk 5]
```diff
@@ -357,11 +371,11 @@ bool EDPServer::processLocalWriterProxyData(
...
-        CacheChange_t* change = writer->first->new_change([]() -> uint32_t
-            {
-                return DISCOVERY_PUBLICATION_DATA_MAX_SIZE;
-            },
-            ALIVE, wdata->key());
+        CacheChange_t* change = writer->first->new_change([this]() -> uint32_t
+                    {
+                        return mp_PDP->builtin_attributes().writerPayloadSize;
+                    },
+                        ALIVE, wdata->key());
```
Comment (IkerLuengo) [inline src/cpp/rtps/builtin/discovery/endpoint/EDPServer.cpp]:
```suggestion
        CacheChange_t* change = writer->first->new_change(
            [this]() -> uint32_t
            {
                return mp_PDP->builtin_attributes().writerPayloadSize;
            },
            ALIVE, wdata->key());
```

File: src/cpp/rtps/builtin/discovery/endpoint/EDPServer.cpp [hunk 6]
```diff
@@ -410,11 +424,11 @@ bool EDPServer::processLocalReaderProxyData(
...
-        CacheChange_t* change = writer->first->new_change([]() -> uint32_t
-            {
-                return DISCOVERY_SUBSCRIPTION_DATA_MAX_SIZE;
-            },
-            ALIVE, rdata->key());
+        CacheChange_t* change = writer->first->new_change([this]() -> uint32_t
+                    {
+                        return mp_PDP->builtin_attributes().writerPayloadSize;
+                    },
+                        ALIVE, rdata->key());
```
Comment (IkerLuengo) [inline src/cpp/rtps/builtin/discovery/endpoint/EDPServer.cpp]:
```suggestion
        CacheChange_t* change = writer->first->new_change(
            [this]() -> uint32_t
            {
                return mp_PDP->builtin_attributes().writerPayloadSize;
            },
            ALIVE, rdata->key());
```

File: src/cpp/rtps/builtin/discovery/endpoint/EDPSimple.cpp [hunk 7]
```diff
@@ -449,32 +453,37 @@ bool EDPSimple::create_sedp_secure_endpoints()
...
-    if(writer->first != nullptr)
+    if (writer->first != nullptr)
     {
         // TODO(Ricardo) Write a getCdrSerializedPayload for ReaderProxyData.
-        CacheChange_t* change = writer->first->new_change([]() -> uint32_t {return DISCOVERY_SUBSCRIPTION_DATA_MAX_SIZE;},
-                ALIVE,rdata->key());
+        CacheChange_t* change = writer->first->new_change([this]() -> uint32_t {
+                        return mp_PDP->builtin_attributes().writerPayloadSize;
+                    },
+                        ALIVE, rdata->key());
```
Comment (IkerLuengo) [inline src/cpp/rtps/builtin/discovery/endpoint/EDPSimple.cpp]:
```suggestion
        CacheChange_t* change = writer->first->new_change(
            [this]() -> uint32_t
            {
                return mp_PDP->builtin_attributes().writerPayloadSize;
            },
            ALIVE, rdata->key());
```

File: src/cpp/rtps/builtin/discovery/endpoint/EDPSimple.cpp [hunk 8]
```diff
@@ -512,25 +521,29 @@ bool EDPSimple::processLocalReaderProxyData(RTPSReader* local_reader, ReaderProx
...
-    if(writer->first !=nullptr)
+    if (writer->first != nullptr)
     {
-        CacheChange_t* change = writer->first->new_change([]() -> uint32_t {return DISCOVERY_PUBLICATION_DATA_MAX_SIZE;},
-                ALIVE, wdata->key());
-        if(change != nullptr)
+        CacheChange_t* change = writer->first->new_change([this]() -> uint32_t {
+                        return mp_PDP->builtin_attributes().writerPayloadSize;
+                    },
+                        ALIVE, wdata->key());
```
Comment (IkerLuengo) [inline src/cpp/rtps/builtin/discovery/endpoint/EDPSimple.cpp]:
```suggestion
        CacheChange_t* change = writer->first->new_change(
            [this]() -> uint32_t
            {
                return mp_PDP->builtin_attributes().writerPayloadSize;
            },
            ALIVE, wdata->key());
```

File: src/cpp/rtps/builtin/discovery/endpoint/EDPSimple.cpp [hunk 9]
```diff
@@ -566,32 +579,35 @@ bool EDPSimple::processLocalWriterProxyData(RTPSWriter* local_writer, WriterProx
...
     {
         InstanceHandle_t iH;
         iH = W->getGuid();
-        CacheChange_t* change = writer->first->new_change([]() -> uint32_t {return DISCOVERY_PUBLICATION_DATA_MAX_SIZE;},
-                NOT_ALIVE_DISPOSED_UNREGISTERED,iH);
-        if(change != nullptr)
+        CacheChange_t* change = writer->first->new_change([this]() -> uint32_t {
+                        return mp_PDP->builtin_attributes().writerPayloadSize;
+                    },
+                        NOT_ALIVE_DISPOSED_UNREGISTERED, iH);
```
Comment (IkerLuengo) [inline src/cpp/rtps/builtin/discovery/endpoint/EDPSimple.cpp]:
```suggestion
        CacheChange_t* change = writer->first->new_change(
            [this]() -> uint32_t
            {
                return mp_PDP->builtin_attributes().writerPayloadSize;
            },
            NOT_ALIVE_DISPOSED_UNREGISTERED, iH);
```

File: src/cpp/rtps/builtin/discovery/endpoint/EDPSimple.cpp [hunk 10]
```diff
@@ -606,32 +622,35 @@ bool EDPSimple::removeLocalWriter(RTPSWriter* W)
...
     {
         InstanceHandle_t iH;
         iH = (R->getGuid());
-        CacheChange_t* change = writer->first->new_change([]() -> uint32_t {return DISCOVERY_SUBSCRIPTION_DATA_MAX_SIZE;},
-                NOT_ALIVE_DISPOSED_UNREGISTERED,iH);
-        if(change != nullptr)
+        CacheChange_t* change = writer->first->new_change([this]() -> uint32_t {
+                        return mp_PDP->builtin_attributes().writerPayloadSize;
+                    },
+                        NOT_ALIVE_DISPOSED_UNREGISTERED, iH);
```
Comment (IkerLuengo) [inline src/cpp/rtps/builtin/discovery/endpoint/EDPSimple.cpp]:
```suggestion
        CacheChange_t* change = writer->first->new_change(
            [this]() -> uint32_t
            {
                return mp_PDP->builtin_attributes().writerPayloadSize;
            },
            NOT_ALIVE_DISPOSED_UNREGISTERED, iH);
```

File: src/cpp/rtps/builtin/discovery/participant/PDP.cpp [hunk 11]
```diff
@@ -377,33 +382,35 @@ bool PDP::enable()
...
-                    return DISCOVERY_PARTICIPANT_DATA_MAX_SIZE;
-                }
-            , ALIVE, key);
+            }
+            // TODO(Ricardo) Change mp_builtin->m_att.writerPayloadSize with getLocalParticipantProxyData()->size().
+            change = mp_PDPWriter->new_change([this]() -> uint32_t
+                        {
+                            return mp_builtin->m_att.writerPayloadSize;
+                        }
+                            , ALIVE, key);
```
Comment (IkerLuengo) [inline src/cpp/rtps/builtin/discovery/participant/PDP.cpp]:
```suggestion
            change = mp_PDPWriter->new_change(
                [this]() -> uint32_t
                {
                    return mp_builtin->m_att.writerPayloadSize;
                }
                , ALIVE, key);
```

File: src/cpp/rtps/builtin/discovery/participant/PDP.cpp [hunk 12]
```diff
@@ -435,15 +442,17 @@ void PDP::announceParticipantState(
...
-            {
-                return DISCOVERY_PARTICIPANT_DATA_MAX_SIZE;
-            }
-        , NOT_ALIVE_DISPOSED_UNREGISTERED, getLocalParticipantProxyData()->m_key);
+        }
+        change = mp_PDPWriter->new_change([this]() -> uint32_t
+                    {
+                        return mp_builtin->m_att.writerPayloadSize;
+                    }
+                        , NOT_ALIVE_DISPOSED_UNREGISTERED, getLocalParticipantProxyData()->m_key);
```
Comment (IkerLuengo) [inline src/cpp/rtps/builtin/discovery/participant/PDP.cpp]:
```suggestion
        change = mp_PDPWriter->new_change(
            [this]() -> uint32_t
            {
                return mp_builtin->m_att.writerPayloadSize;
            }
            , NOT_ALIVE_DISPOSED_UNREGISTERED, getLocalParticipantProxyData()->m_key);
```

File: src/cpp/rtps/builtin/discovery/participant/PDPClient.cpp [hunk 13]
```diff
@@ -466,13 +475,15 @@ void PDPClient::announceParticipantState(
...
         assert(pW);
 
         CacheChange_t* change = nullptr;
 
-        if ((change = pW->new_change([]() -> uint32_t {return DISCOVERY_PARTICIPANT_DATA_MAX_SIZE; },
-            NOT_ALIVE_DISPOSED_UNREGISTERED, getLocalParticipantProxyData()->m_key)))
+        if ((change = pW->new_change([this]() -> uint32_t {
+                        return mp_builtin->m_att.writerPayloadSize;
+                    },
+                NOT_ALIVE_DISPOSED_UNREGISTERED, getLocalParticipantProxyData()->m_key)))
```
Comment (IkerLuengo) [inline src/cpp/rtps/builtin/discovery/participant/PDPClient.cpp]:
```suggestion
        if ((change = pW->new_change(
            [this]() -> uint32_t
            {
                return mp_builtin->m_att.writerPayloadSize;
            },
            NOT_ALIVE_DISPOSED_UNREGISTERED, getLocalParticipantProxyData()->m_key)))
```

File: src/cpp/rtps/builtin/discovery/participant/PDPServer.cpp [hunk 14]
```diff
@@ -543,22 +562,25 @@ bool PDPServer::trimPDPWriterHistory()
...
     if (sid == SampleIdentity::unknown())
     {
         sid.writer_guid(c.writerGUID);
         sid.sequence_number(c.sequenceNumber);
-        logError(RTPS_PDP, "A DATA(p) received by server " << mp_PDPWriter->getGuid()
-            << " from participant " << c.writerGUID << " without a valid SampleIdentity");
+        logError(RTPS_PDP,
+                "A DATA(p) received by server " << mp_PDPWriter->getGuid()
+                                                << " from participant " << c.writerGUID <<
+                            " without a valid SampleIdentity");
```
Comment (IkerLuengo) [inline src/cpp/rtps/builtin/discovery/participant/PDPServer.cpp]:
```suggestion
        logError(RTPS_PDP,
                "A DATA(p) received by server " << mp_PDPWriter->getGuid()
                    << " from participant " << c.writerGUID
                    << " without a valid SampleIdentity");
```

File: src/cpp/rtps/builtin/discovery/participant/PDPServer.cpp [hunk 15]
```diff
@@ -753,8 +779,10 @@ void PDPServer::announceParticipantState(bool new_change, bool dispose /* = fals
 
             CacheChange_t* change = nullptr;
 
-            if ((change = pW->new_change([]() -> uint32_t {return DISCOVERY_PARTICIPANT_DATA_MAX_SIZE; },
-                NOT_ALIVE_DISPOSED_UNREGISTERED, getLocalParticipantProxyData()->m_key)))
+            if ((change = pW->new_change([this]() -> uint32_t {
+                            return mp_builtin->m_att.writerPayloadSize;
+                        },
+                    NOT_ALIVE_DISPOSED_UNREGISTERED, getLocalParticipantProxyData()->m_key)))
```
Comment (IkerLuengo) [inline src/cpp/rtps/builtin/discovery/participant/PDPServer.cpp]:
```suggestion
            if ((change = pW->new_change(
                [this]() -> uint32_t
                {
                    return mp_builtin->m_att.writerPayloadSize;
                },
                NOT_ALIVE_DISPOSED_UNREGISTERED, getLocalParticipantProxyData()->m_key)))
```

File: src/cpp/rtps/builtin/discovery/participant/PDPServer.cpp [hunk 16]
```diff
@@ -891,15 +926,17 @@ bool PDPServer::remove_remote_participant(
...
+        if (!(mp_PDPReaderHistory->get_max_change(&pC) &&
                 pC->kind == NOT_ALIVE_DISPOSED_UNREGISTERED && // last message received is aun DATA(p[UD])
                 pC->instanceHandle == key )) // from the same participant I'm going to report
         {   // We must create the DATA(p[UD])
-            if ((pC = mp_PDPWriter->new_change([]() -> uint32_t {return DISCOVERY_PARTICIPANT_DATA_MAX_SIZE; },
-                NOT_ALIVE_DISPOSED_UNREGISTERED, key)))
+            if ((pC = mp_PDPWriter->new_change([this]() -> uint32_t {
+                            return mp_builtin->m_att.writerPayloadSize;
+                        },
+                    NOT_ALIVE_DISPOSED_UNREGISTERED, key)))
```
Comment (IkerLuengo) [inline src/cpp/rtps/builtin/discovery/participant/PDPServer.cpp]:
```suggestion
            if ((pC = mp_PDPWriter->new_change(
                [this]() -> uint32_t
                {
                    return mp_builtin->m_att.writerPayloadSize;
                },
                NOT_ALIVE_DISPOSED_UNREGISTERED, key)))
```

File: src/cpp/rtps/builtin/discovery/endpoint/EDPServer.cpp [hunk 17]
```diff
@@ -232,13 +244,11 @@ bool EDPServer::addEndpointFromHistory(
     // See if this sample is already in the cache.
     // TODO: Accelerate this search by using a PublisherHistory as mp_PDPWriterHistory
     auto it = std::find_if(history.changesRbegin(), history.changesRend(), [&sid](CacheChange_t* c) {
```
Comment (IkerLuengo) [inline src/cpp/rtps/builtin/discovery/endpoint/EDPServer.cpp]:
```suggestion
    auto it = std::find_if(history.changesRbegin(), history.changesRend(),
            [&sid](CacheChange_t* c)
            {
```

File: src/cpp/rtps/builtin/discovery/endpoint/EDPServer.cpp [hunk 18]
```diff
@@ -232,13 +244,11 @@ bool EDPServer::addEndpointFromHistory(
...
-        return sid == c->write_params.sample_identity();
-    });
+                    return sid == c->write_params.sample_identity();
+                });
```
Comment (IkerLuengo) [inline src/cpp/rtps/builtin/discovery/endpoint/EDPServer.cpp]:
```suggestion
                return sid == c->write_params.sample_identity();
            });
```

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1505/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1276/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1503/)

**Pull request merged**